### DISCUSSION TITLE
undefined variable object.dataSource

### DISCUSSION THREAD

**Discussion by LuckyPickleZZ** [dbeaver/dbeaver#13353] [DBeaver Development] [answered]
#### System information: 
- Operating system (distribution) and version 
   OS=macosx, ARCH=x86_64, WS=cocoa, NL=zh_CN_#Hans
- DBeaver version
  org.jkiss.dbeaver.core 21.1.3.qualifier
- Additional extensions
  Custom plugin

#### Describe the problem you're observing:
  When I put visible if in the first folder tag, an error will be reported. But not in the back. How should this be solved?
```
<tree
    icon="icons/ob_icon.png"
    label="OceanBase data source"
    path="oceanbase">
        <folder
            type="org.jkiss.dbeaver.ext.oceanbase.mysql.model.OceanbaseMySQLCatalog"
            label="%tree.databases.node.name"
            description="%tree.databases.node.tip"
            visibleIf="object.dataSource.isMySQLMode()">
        </folder>
</tree>
```

The error is
```
org.apache.commons.jexl3.JexlException$Variable: org.jkiss.dbeaver.model.impl.AbstractDescriptor.parseExpression@1:2 undefined variable object.dataSource
```

**Comment by ShadelessFox** [comment, marked as answer]
Hello @LuckyPickleZZ,

Data source is not available at this point.

Navigator is trying to determine whether connection has any children (in your case it's `OceanbaseMySQLCatalog`) by evaluating your expression, but connection is not opened and thus no data source is present.

**Reply by LuckyPickleZZ** [reply]
Thanks a lot. What you said is the same as I guessed. But if I want to control the display of this folder tag at present, what should I do ?
I had also tried to define a method directly in the `OceanbaseMySQLCatalog` without obtaining the datasource. However, this method is also unsuccessful and will report `unresolved variable object`.

**Reply by ShadelessFox** [reply]
In order to get `OceanbaseMySQLCatalog`, you have to obtain data source first, so this will not work.

You can keep this folder always visible and hide its children if `isMySQLMode` is not set. I think this solution is more appropriate.

**Reply by LuckyPickleZZ** [reply]
All right.. In that case, I'll keep it visible. Thank you very very much~

**Reply by ShadelessFox** [reply]
You're welcome!

**Comment by ShadelessFox** [comment]
You can take look at [DBXTreeNode#makeContext](https://github.com/dbeaver/dbeaver/blob/devel/plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/meta/DBXTreeNode.java#L344-L370) and extend it to fit your needs.
